doc/reference/trd-hil-design.md

```diff
@@ -0,0 +1,685 @@
+**TRD:** <br/>
+**Working Group:** Kernel<br/>
+**Type:** Best Current Practice<br/>
+**Status:** Draft <br/>
+**Author:** Philip Levis <br/>
+**Draft-Created:** April 1, 2021<br/>
+**Draft-Modified:** April 9, 2021<br/>
+**Draft-Version:** 3<br/>
+**Draft-Discuss:** tock-dev@googlegroups.com</br>
+
+Abstract
+-------------------------------
+
+This document describes design rules of hardware interface layers (HILs)
+in the Tock operating system. HILs are Rust traits that provide a
+standard interface to a hardware resource, such as a sensor, a flash
+chip, a cryptographic accelerator, a bus, or a radio. Developers
+adding new HILs to Tock should read this document and verify they have
+followed these guidelines.
+
+Introduction
+===============================
+
+In Tock, a hardware interface layer (HIL) is a collection of Rust
+traits and types that provide a standardized API to a hardware
+resource such as a sensor, flash chip, cryptographic accelerator, bus,
+or a radio. Capsules typically use HILs to provide their
+functionality. For example, a system call driver capsule that gives
+processes access to a temperature sensor relies on having a reference
+to an implementation of the `kernel::hil::sensors::TemperatureDriver`
+trait. This allows the system call driver capsule to work on top of
+any implementation of the `TemperatureDriver` trait, whether it is a
+local, on-chip sensor, an analog sensor connected to an ADC, or a
+digital sensor over a bus.
+
+HILs are used for many purposes within the kernel. They can be
+directly accessed by kernel services, such as the in-kernel process
+console using the UART HIL. They can be exposed to processes with
+system driver capsules, such as with GPIO. They can be virtualized to
+allow multiple clients to share a single resource, such as with the
+virtual timer capsule.
+
+This variety of use cases places a complex set of requirements on how a
+HIL must behave. For example, Tock expects that every HIL is
+virtualizable: it is possible to take one instance of the trait and
+allow multiple clients to use it simultaneously, such that each one
+thinks it has its own, independent instance of the trait. Because
+virtualization often means requests can be queued and the Tock kernel
+has a single stack, all HILs must be nonblocking and so have a
+callback for completion. This has implications to buffer management
+and ownership.
+
+This document describes these requirements and describes a set of
+design rules for HILs. They are:
+
+1. Do not issue synchronous callbacks.
+2. Split-phase operations return a synchronous `Result` type which 
+   includes an error code in its `Err` value.
+3. For split-phase operations, `Ok` means a callback will occur
+   while `Err` means one won't.
+4. Error results of split-phase operations with a buffer parameter 
+   include a reference to passed buffer. This returns the buffer 
+   to the caller.
+5. Split-phase operrations with a buffer parameter take a mutable reference 
```

Comment:
"operations" is typo'd here.

Reply:
Haha, I already suggested that above but GitHub has conveniently hidden it. :laughing:

Reply:
Has been fixed in 730b2dbcda0c00a977fe187fd23d921cbd141550.